### PR TITLE
add API endpoints to update/delete protected branch

### LINKED ISSUE #2922: API endpoint to configure default branch and protected branches is missing
- Gitea version: 1.2.3

## Description
Gitea allows to configure the default branch and protected branches via GITEA_URL/org/repo/settings/branches but I could not find an API alternative in the local Swagger documentation.

### PR DESCRIPTION
**Pull request by quantonganh**
Fix https://github.com/go-gitea/gitea/issues/2922

### CODE REVIEW

**Comment by techknowlogick** [comment]
Thanks for PR :) The CI is currently failing due to needing a rebase on the latest commit of master.

I think additional routes for the API are needing for these methods, as well as swagger definitions.

**Comment by quantonganh** [comment]
> I think additional routes for the API are needing for these methods, as well as swagger definitions.

Routes and swagger definitions have been added.
@techknowlogick Could you please take a review?

File: routers/api/v1/repo/branch.go [hunk 1]
```diff
@@ -88,19 +101,204 @@ func ListBranches(ctx *context.APIContext) {
...
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/Branch"
+	protectBranch, err := models.GetProtectedBranchBy(ctx.Repo.Repository.ID, ctx.Params(":branch"))
+	if err != nil {
+		if !git.IsErrBranchNotExist(err) {
+			ctx.Error(http.StatusInternalServerError, "GetProtectedBranchBy", err)
 			return
 		}
```
Comment (davidsvantesson) [inline routers/api/v1/repo/branch.go:158]: Xorm will not give an error if there is no row in the database, but an empty object. 
You need to handle:
- If there is no protected branch, but the branch exist (create protectBranch with default values / protection:false )
- If the branch doesn't exist. (404)

File: routers/api/v1/repo/branch.go [hunk 2]
```diff
@@ -88,19 +101,204 @@ func ListBranches(ctx *context.APIContext) {
...
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/Branch"
+	branch := ctx.Params(":branch")
+	protectBranch, err := models.GetProtectedBranchBy(ctx.Repo.Repository.ID, branch)
+	if err != nil {
+		if !git.IsErrBranchNotExist(err) {
+			ctx.Error(http.StatusInternalServerError, "GetProtectedBranchBy", err)
+			return
+		}
```
Comment (davidsvantesson) [inline routers/api/v1/repo/branch.go:203]: Same problem as above with xorm.
If the protected branch does not exist, you need to check if the branch exist (or give 404)

File: routers/api/v1/repo/branch.go [hunk 3]
```diff
@@ -88,19 +101,204 @@ func ListBranches(ctx *context.APIContext) {
...
+	if !protected {
+		ctx.Status(http.StatusNoContent)
+		return
+	}
+	protectBranch, err := models.GetProtectedBranchBy(ctx.Repo.Repository.ID, branchName)
+	if err != nil {
+		if !git.IsErrBranchNotExist(err) {
+			ctx.Error(http.StatusInternalServerError, "GetProtectedBranchBy", err)
+			return
+		}
```
Comment (davidsvantesson) [inline routers/api/v1/repo/branch.go:297]: Same problem as above with xorm. Maybe ok to give 404 directly if protected branch doesn't exist.

**Review by davidsvantesson — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/branch.go (hunk 1 above)
Comment (davidsvantesson) [reply routers/api/v1/repo/branch.go]: Also, xorm will not give an error if there is no row in the database, but an empty object. So you have to check protectBranch against nil.

File: templates/swagger/v1_json.tmpl [hunk 4]
```diff
@@ -9115,6 +9243,44 @@
       },
       "x-go-package": "code.gitea.io/gitea/modules/structs"
     },
+    "ProtectBranchForm": {
```
Comment (davidsvantesson) [inline templates/swagger/v1_json.tmpl:9703]: It's a bit problematic to reuse the same structure for API from documentation point of view. How do I know the format of XWhiteListTeams and XWhiteListUsers? Preferably they should be arrays. At least they should change name to XWhiteListTeamIDs and XWhiteListUserIDs

File: modules/structs/repo_branch.go [hunk 5]
```diff
@@ -6,6 +6,7 @@ package structs
 
 // Branch represents a repository branch
 type Branch struct {
-	Name   string         `json:"name"`
-	Commit *PayloadCommit `json:"commit"`
+	Name      string         `json:"name"`
+	Commit    *PayloadCommit `json:"commit"`
+	Protected bool           `json:"protected"`
 }
```
Comment (lunny) [inline modules/structs/repo_branch.go:12]: I think we need more details about the branch protection settings but not only a `bool`.

**Comment by davidsvantesson** [comment]
I think this one can be moved to 1.12, no chance for this release. I hope either this one or my new PR can get into next release.

**Comment by davidsvantesson** [comment]
This one can be closed.

**Pull request closed without merging**